tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py

```diff
@@ -281,12 +287,77 @@ def _zeros_slot(var, slot_name, op_name):
       named_slots[optimizer._var_key(var)] = new_slot_variable
     return named_slots[optimizer._var_key(var)]
 
+  def apply_gradients(grads_and_vars, global_step=None, name=None):
+    """Apply gradients to variables.
+    Args:
+      grads_and_vars: List of (gradient, variable) pairs as returned by
+        compute_gradients().
+      global_step: Optional Variable to increment by one after the
+        variables have been updated.
+      name: Optional name for the returned operation.  Default to the
+        name passed to the Optimizer constructor.
+
+    Returns:
+      train_op: apply gradients op to be executed by each replica.
+
+    Raises:
+      ValueError: If the grads_and_vars is empty.
+      ValueError: If global step is not provided, the staleness cannot be
+        checked.
+    """
+    try:
+      import horovod.tensorflow as hvd
+    except ImportError:
+      raise ValueError(
+          "Please install Horovod first if you want to use distributed synchronous training ased on Horovod"
+      )
+    if not grads_and_vars:
+      raise ValueError("Must supply at least one variable")
+
+    if global_step is None:
+      raise ValueError("Global step is required to check staleness")
+
+    trainable_grad_and_vars = []
+    aggregated_grad = []
+    var_list = []
+
+    with backend.name_scope(name or self._name):
+      for grad, var in grads_and_vars:
+        if isinstance(var, de.TrainableWrapper):
+          trainable_grad_and_vars.append((grad, var))
+          continue
+        var_list.append(var)
+        with ops.device(var.device):
+          # Dense gradients.
+          if grad is None:
+            aggregated_grad.append(None)  # pass-through.
+            continue
+          elif isinstance(grad, ops.Tensor):
```

Comment:
MLP layers, downstream from embedding, may also generate IndexedSlices gradient, need to notice that.

Reply:
Sorry, I never notice MLP layers  could generate IndexedSlices gradient, could you give me an example?

Reply:
```

Reply:
> Sorry, I never notice MLP layers could generate IndexedSlices gradient, could you give me an example?

For example:

```python
var = tf.Variable(...)
params = de.get_variable(...)
emb = de.embedding_lookup(...)
latent_tensor = sum_pooling(emb)
... = some_func(latent_tensor, var)
```

The `some_layer` are defined as:
```python
def some_func(latent, var):
  mask = tf.greater_equal(var, threshold)
  pos = tf.where(mask)
  selected = tf.gather(var, pos)
  return tf.reduce_mean(selected)
```

The `some_func` code will make the gradient become IndexedSlices to `var`.

Reply:
I see, Thanks. Now I delelte the judgement `isinstance(grad, ops.Tensor)`. So Horovod will check it's grad type and deal with it.